src/photos/ducks/clustering/albums.js

```diff
@@ -91,54 +123,80 @@ const createClusters = async clusters => {
   return refsCount
 }
 
-const removeClusters = async (clusters, albumsToSave) => {
-  let refsCount = 0
-  for (const photos of clusters) {
-    refsCount += await removeAutoAlbumsReferences(photos, albumsToSave)
-  }
-  await removeAutoAlbums(albumsToSave)
-  return refsCount
-}
-
-export const saveClustering = async (clusters, albumsToSave) => {
+/**
+ * Save the clusters in database. Each cluster consists of a set of photos
+ * that must be referenced-by the same auto-album.
+ * The clusterAlbums are the existing auto-albums that must be either updated
+ * or deleted.
+ *
+ * @param {Object[]} clusters - Set of cluster of photos
+ * @param {Object[]} clusterAlbums - Set of existing auto-albums
+ * @returns {number} Number of references updated in database
+ *
+ */
+export const saveClustering = async (clusters, clusterAlbums) => {
   let refsCount = 0
-  if (albumsToSave && albumsToSave.length > 0) {
-    if (albumsToSave.length > 1 || clusters.length > 1) {
-      // The clustering structure has changed: remove the impacted clusters
-      // and create the new ones
-      const refsRemoved = await removeClusters(clusters, albumsToSave)
-      const refsAdded = await createClusters(clusters, albumsToSave)
-      refsCount = refsCount + refsAdded - refsRemoved
-    } else {
-      // The clustering structure has not changed: only new photos to add
-      refsCount = flatten(
-        await Promise.all(
-          clusters.map((photos, index) => {
-            return addAutoAlbumReferences(photos, albumsToSave[index])
-          })
-        )
-      ).reduce((acc, val) => acc + val, 0)
+  if (clusterAlbums && clusterAlbums.length > 0) {
+    const processedAlbumsIds = []
+    for (const photos of clusters) {
+      // Find the clusterIds for this set of photos
+      const clusterIds = uniq(
+        photos.filter(p => p.clusterId).map(p => p.clusterId)
+      )
+      if (clusterIds.length === 0) {
+        // No clusterId : create the new cluster
+        const album = await createAutoAlbum(photos)
+        refsCount += await addAutoAlbumReferences(photos, album)
+      } else if (clusterIds.length === 1) {
+        const album = clusterAlbums.find(album => album._id === clusterIds[0])
+        if (processedAlbumsIds.includes(album._id)) {
+          // Album already processed for another cluster: remove the refs and create a new album
```

Comment:
Consider this scenario: we have 2 clusters: c1 = [p1, p2], c2 = [p3]. p1 and p3 used to be in the same cluster, so referenced by the same album, say a1. p2 is a newly uploaded photo.
We first process c1. a1 is found because of p1,so we add it as a `referenced-by` in p2.
Then we process c2. a1 is found again, but it has already been used in c1, another cluster. So we cannot use it anymore: we remove the reference and create a3, the new `referenced-by` of p3.